framework/src/userobject/UserObject.C

```diff
@@ -36,12 +36,18 @@ validParams<UserObject>()
                         "in the case this is true but no "
                         "displacements are provided in the Mesh block "
                         "the undisplaced mesh will still be used.");
-  params.addParamNamesToGroup("use_displaced_mesh", "Advanced");
+  params.addParam<bool>("allow_duplicate_execution_on_initial",
+                        false,
+                        "In the case where this UserObject is depended upon by an initial "
+                        "condition, allow it to be executed twice during the initial setup (once "
+                        "before the IC and again after mesh adaptivity (if applicable).");
```

Comment:
It doesn't need to re-execute on each initial adaptivity step. The adaptivity steps will project the initial condition each time to the new mesh. When that's all done, we just need to rerun it once on the new mesh and deal with the changes. That's what's being done now. Most of the grain growth examples have 2 adaptivity steps, and I have examined the results and everything looks exactly correct.